File: leetcode/LinkedListTests/LinkedList_1472.cs
namespace LinkedListTests;

[TestFixture]
class LinkedList_1472
{
    public class BrowserHistory {
        private DoubleNode history;

        public BrowserHistory(string homepage) {
            history = new DoubleNode(homepage);
        }
    
        public void Visit(string url) {
            history.next = new DoubleNode(url);
            history.next.prev = history;
            history = history.next;
        }
    
        public string Back(int steps) {
            while(history.prev is not null && steps >0) {
                history = history.prev;
                steps--;
            }
            return history.val;
        }
    
        public string Forward(int steps) {
            while(history.next is not null && steps >0) {
                history = history.next;
                steps--;
            }
            return history.val;
        }
    }
}

File: leetcodePython/LinkedList/linkedlist_1472.py
class Node:
    def __init__(self, url: str):
        self.url = url
        self.next = None
        self.prev = None


class BrowserHistory:
    def __init__(self, homepage: str):
        self.current = Node(homepage)

    def visit(self, url: str) -> None:
        self.current.next = Node(url)
        self.current.next.prev = self.current
        self.current = self.current.next

    def back(self, steps: int) -> str:
        for i in range(steps):
            if self.current.prev is None:
                break
            self.current = self.current.prev
        return self.current.url

    def forward(self, steps: int) -> str:
        for i in range(steps):
            if self.current.next is None:
                break
            self.current = self.current.next
        return self.current.url
